engine/runtime-language-arrow/src/main/java/org/enso/interpreter/arrow/runtime/ArrowFixedSizeArrayBuilder.java

```diff
@@ -0,0 +1,353 @@
+package org.enso.interpreter.arrow.runtime;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.ImportStatic;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.interop.*;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import java.time.*;
+import org.enso.interpreter.arrow.LogicalLayout;
+
+@ExportLibrary(InteropLibrary.class)
+public final class ArrowFixedSizeArrayBuilder implements TruffleObject {
+  private final ByteBufferDirect buffer;
+  private final LogicalLayout unit;
+  private final int size;
+  private int index;
+
+  private boolean sealed;
+
+  public ArrowFixedSizeArrayBuilder(int size, LogicalLayout unit) {
+    this.size = size;
+    this.unit = unit;
+    this.buffer = ByteBufferDirect.forSize(size, unit);
+    this.index = 0;
+    this.sealed = false;
+  }
+
+  public LogicalLayout getUnit() {
+    return unit;
+  }
+
+  @ExportMessage
+  public boolean hasArrayElements() {
+    return true;
```

Comment:
The _"builder"_ is in fact also an (Enso) `Array` and it can be mutated until somebody calls `build` on it. I'd be interested to see a _use-cases_ for such a design decision...